Fixes in persistent error log

gamanakis opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
While testing PR #12355 I came across some bugs in persistent error log.

### Description
<!--- Describe your changes in detail -->
Address the following bugs in persistent error log:
1) Check nested clones, eg "fs->snap->clone->snap2->clone2".

2) When deleting files containing error blocks in those clones (from
   "clone" the example above), do not break the check chain.

3) When deleting files in the originating fs before syncing the errlog
   to disk, do not break the check chain. This happens because at the
   time of introducing the error block in the error list, we do not have
   its birth txg and the head filesystem. If the original file is
   deleted before the error list is synced to the error log (which is
   when we actually lookup the birth txg and the head filesystem), then
   we do not have access to this info anymore and break the check chain.

The most prominent change is related to achieving (3). We expand the
spa_error_entry_t structure to accomodate the newly introduced
zbookmark_err_phys_t structure (containing the birth txg of the error
block).Due to compatibility reasons we cannot remove the
zbookmark_phys_t structure and we also need to place the new structure
after se_avl, so it is not accounted for in avl_find(). Then we modify
spa_log_error() to also provide the birth txg of the error block. With
these changes in place we simplify the previously introduced function
get_head_and_birth_txg() (now named get_head_ds()).

We chose not to follow the same approach for the head filesystem (thus
completely removing get_head_ds()) to avoid introducing new lock
contentions.

The stack sizes of nested functions (as measured by checkstack.pl in the
linux kernel) are:
check_filesystem [zfs]: 272 (was 912)
check_clones [zfs]: 64

We also introduced two new tests covering the above changes.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### How Has This Been Tested?
Added tests in ZTS.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf commented:
This 22.04 functional (part 1) failure was a CI issue.  If you rebase after c9c463a508bf11807dbf3b5919b9cb8c2778518d that should fix the issue.

gamanakis commented:
There seems to be some lock contention going on, will take a look.

behlendorf commented on include/sys/spa_impl.h:
```diff
@@ -66,6 +66,7 @@ typedef struct spa_error_entry {
 	zbookmark_phys_t	se_bookmark;
 	char			*se_name;
 	avl_node_t		se_avl;
+	zbookmark_err_phys_t	se_zep;		/* not accounted in avl_find */
```
By not accounted for you simply mean it's not used by the `spa_error_entry_compare` function during `avl_find()`.  Right?

gamanakis replied:
Right. Thank you for taking a look!

gamanakis replied:
I did it that way due to compatibility reasons.

behlendorf reviewed (approved):
Nice work, I'm glad you caught this now.  Stashing away the birth txg immediately in `spa_log_error()` when we still have it makes a lot of sense.  And it was nice that we could largely avoid any compatibility issues.

behlendorf commented:
@gamanakis if you can fix up the FreeBSD build error introduced with this change I think we can get it merged.

http://build.zfsonlinux.org/builders/FreeBSD%20stable%2F13%20amd64%20%28TEST%29/builds/8120

```
In file included from /var/lib/buildbot/slaves/zfs/FreeBSD_stable_13_amd64__TEST_/build/zfs/module/icp/algs/blake3/blake3_impl.c:284:
In file included from /var/lib/buildbot/slaves/zfs/FreeBSD_stable_13_amd64__TEST_/build/zfs/module/icp/include/generic_impl.c:37:
In file included from /var/lib/buildbot/slaves/zfs/FreeBSD_stable_13_amd64__TEST_/build/zfs/include/sys/zio_checksum.h:31:
/var/lib/buildbot/slaves/zfs/FreeBSD_stable_13_amd64__TEST_/build/zfs/include/sys/zio.h:311:3: error: redefinition of typedef 'zbookmark_err_phys_t' is a C11 feature [-Werror,-Wtypedef-redefinition]
} zbookmark_err_phys_t;
  ^
/var/lib/buildbot/slaves/zfs/FreeBSD_stable_13_amd64__TEST_/build/zfs/include/sys/spa.h:68:35: note: previous definition is here
typedef struct zbookmark_err_phys zbookmark_err_phys_t;
                                  ^
1 error generated.
*** [blake3_impl.o] Error code 1
```

gamanakis commented:
https://github.com/openzfs/zfs/commit/70e185fee822047904c544af791f061fd8657331: Rebased to master, removed the surplus definition of `zbookmark_err_phys_t` in spa.h.

gamanakis commented:
[32fc496](https://github.com/openzfs/zfs/commit/32fc496a81ad50275dca10e548835377179632f8): When defining `zbookmark_err_phys_t` follow the paradigm of `zbookmark_phys_t`. I think this is more suitable.
